validation/firefox-accounts/amplitude-event.1.fxa-payments-server--fxa_pay_setup--engage.pass.json

```diff
@@ -0,0 +1,12 @@
+{
+  "op": "amplitudeEvent",
+  "event_type": "fxa_pay_setup - engage",
+  "time": 1584621762045,
+  "os_name": "Windows",
+  "app_version": "162.1",
+  "os_version": "10",
+  "user_properties": "{\"flow_id\":\"cdb4d564b76249355c8d9e8d0741db2b192f4cd7185ef5a1c197c2cf06d32610\",\"ua_browser\":\"Firefox\",\"ua_version\":\"72.0\"}",
+  "device_id": "61e8366b8d374ad08f49b59c79904bab",
+  "session_id": "1584261684526",
+  "event_properties": "{\"plan_id\":\"plan_FiJ55YK6UYftPa\",\"product_id\":\"prod_FiJ42WCzZNRSbS\"}"
```

Comment:
Do you know the history of why event_properties and user_properties are encoded as strings containing json documents rather than as non-stringified json objects? Is there any possibility of changing that format in what FxA sends to the pipeline?

This isn't a blocker for getting these initial schemas merged, but just be aware that we essentially won't be able to do any validation on the contents of these fields if they're in string form.

Reply:
Great catch! Now that you mention it, this looks like a bug in the FxA server metrics code, because there actually is [a spot](https://github.com/mozilla/fxa-amplitude-send/blob/89bfaef20a1d978fce2dccddc0155699a17ac172/pubsub/utils.js#L40-L46) in the existing `fxa-amplitude-send` message parsing code that detects these double-serialized fields and double-parses them:

```
function parseMessage (event, HMAC_KEY) {
  if (event.Fields) {
    event = event.Fields

    if (event.op && event.data) {
      event = JSON.parse(event.data)
    } else {
      if (is.nonEmptyString(event.event_properties)) {
        event.event_properties = JSON.parse(event.event_properties)
      }

      if (is.nonEmptyString(event.user_properties)) {
        event.user_properties = JSON.parse(event.user_properties)
      }
    }
  }
```

I've filed a bug to investigate why this is happening: https://github.com/mozilla/fxa/issues/4611